util/oci/client.go

```diff
@@ -0,0 +1,367 @@
+	return NewClientWithLock(repoURL, creds, globalLock, proxy, opts...)
+}
+func NewClientWithLock(repoURL string, creds Creds, repoLock sync.KeyLock, proxyUrl string, opts ...ClientOpts) (Client, error) {
+	ociRepo := strings.TrimPrefix(repoURL, "oci://")
+	repo, err := remote.NewRepository(ociRepo)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize repository: %v", err)
+	}
+
+	repo.PlainHTTP = creds.InsecureHttpOnly
+
+	ociUri, err := url.Parse(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse registry host: %v", err)
+	}
+
+	var tlsConf *tls.Config
+	if !repo.PlainHTTP {
+		tlsConf, err = newTLSConfig(creds)
+		if err != nil {
+			return nil, fmt.Errorf("failed setup tlsConfig: %v", err)
+		}
+	}
+
+	var ociRegistry string
+	if ociUri.Port() != "" {
+		ociRegistry = fmt.Sprintf("%s:%s", ociUri.Hostname(), ociUri.Port())
+	} else {
+		ociRegistry = ociUri.Hostname()
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		Proxy:             proxy.GetCallback(proxyUrl),
+		TLSClientConfig:   tlsConf,
+		DisableKeepAlives: true,
+	}}
+	repo.Client = &auth.Client{
+		Client: client,
+		Cache:  nil,
+		Credential: auth.StaticCredential(ociRegistry, auth.Credential{
+			Username: creds.Username,
+			Password: creds.Password,
+		}),
+	}
+
+	c := &nativeOciClient{
+		creds:    creds,
+		repoURL:  ociRepo,
+		proxy:    proxyUrl,
+		repoLock: repoLock,
+		repo:     repo,
+	}
+	for i := range opts {
+		opts[i](c)
+	}
+	return c, nil
+}
+
+// nativeOciClient implements Client interface using oras-go
+type nativeOciClient struct {
+	creds          Creds
+	repoURL        string
+	proxy          string
+	repo           *remote.Repository
+	repoLock       sync.KeyLock
+	indexCache     indexCache
+	repoCachePaths argoio.TempPaths
+}
+
+func (c *nativeOciClient) TestRepo(ctx context.Context) (bool, error) {
+	err := c.repo.Tags(ctx, "", func(tags []string) error {
+		return nil
+	})
+	return err == nil, err
+}
+
+func (c *nativeOciClient) Extract(ctx context.Context, digest string, project string, manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool) (string, argoio.Closer, error) {
+	cachedPath, err := c.getCachedPath(digest, project)
+	if err != nil {
+		return "", nil, err
+	}
+
+	c.repoLock.Lock(cachedPath)
+	defer c.repoLock.Unlock(cachedPath)
+
+	exists, err := fileExist(cachedPath)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if !exists {
+		// Create scratch dir for the OCI store. This will be put into a TAR archive.
+		tempDir, err := files.CreateTempDir(os.TempDir())
+		if err != nil {
+			return "", nil, err
+		}
+		defer os.RemoveAll(tempDir)
+
+		store, err := oci.New(tempDir)
+		if err != nil {
+			return "", nil, err
+		}
+
+		// Copy remote repo at the given digest to the scratch dir.
+		_, err = oras.Copy(ctx, c.repo, digest, store, digest, oras.DefaultCopyOptions)
+		if err != nil {
+			return "", nil, err
+		}
+
+		err = createTarFile(tempDir, cachedPath)
```

Comment:
@shizhMSFT what do you mean? That I should remove `tempDir`?

Reply:
@blakepettersson You've already done `defer os.RemoveAll(tempDir)`. I was talking about something else.

Let me elaborate more on this. When you do `oras copy`, `oras.Copy()`, or any push operation, `oras` will put temporary files under the `ingest` folder so that the OCI image layout folder (the CAS part) can be read / write by multiple oras process at the same time without race condition.

```console
$ oras cp docker.io/argoproj/argocd:v2.6.15 --to-oci-layout argocd:v2.6.15 2>/dev/null
Copied [registry] docker.io/argoproj/argocd:v2.6.15 => [oci-layout] argocd:v2.6.15
Digest: sha256:58ebb4ed23c8db4cd4cc3f954f8d94c4b4e3d9669c0751c484108d22b86d52de
$ ls argocd
blobs  index.json  ingest  oci-layout
$ ls argocd/ingest
```

As you can observe above, there is an empty folder `ingest`, which was used by `oras cp`. It is not meaningful to include it in a tar file. So... it is better to remove it before tar.

```console
$ rmdir argocd/ingest
$ cd argocd; tar cf ../argocd.tar *; cd ..
$ oras repo tags --oci-layout argocd.tar
v2.6.15
```